README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->
bitflyer
========

[![Build Status](https://travis-ci.org/teramonagi/bitflyer.svg?branch=master)](https://travis-ci.org/teramonagi/bitflyer)

R wrapper for bitFlyer's REST API

Installation
------------

``` r
install.packages("bitflyer")

# Or the development version from GitHub:
# install.packages("devtools")
devtools::install_github("teramonagi/bitflyer")
```

Example
-------

### Public API

``` r
library("bitflyer")
# Get the list of market
get_markets()
#> [1] "[{\"product_code\":\"BTC_JPY\"},{\"product_code\":\"FX_BTC_JPY\"},{\"product_code\":\"ETH_BTC\"},{\"product_code\":\"BCH_BTC\"},{\"product_code\":\"BTCJPY28SEP2018\",\"alias\":\"BTCJPY_MAT3M\"},{\"product_code\":\"BTCJPY20JUL2018\",\"alias\":\"BTCJPY_MAT1WK\"},{\"product_code\":\"BTCJPY27JUL2018\",\"alias\":\"BTCJPY_MAT2WK\"}]"

# Get order book
x <- board(product_code = "BTC_JPY")
stringr::str_sub(x, 1, 200)
#> [1] "{\"mid_price\":835199.0,\"bids\":[{\"price\":835050.0,\"size\":0.49763534},{\"price\":834984.0,\"size\":1.0},{\"price\":834870.0,\"size\":0.02},{\"price\":834836.0,\"size\":0.3},{\"price\":834812.0,\"size\":0.01},{\"price\":83"

# Get the current status of the exchange.
get_health()
#> [1] "{\"status\":\"NORMAL\"}"

# ... etc
```

### Private API

``` r
# Get a list of which HTTP Private APIs can be used with the specified API key
get_permissions()
#> [1] "[\"/v1/getmarkets\",\"/v1/markets\",\"/v1/getmarkets/usa\",\"/v1/markets/usa\",\"/v1/getmarkets/eu\",\"/v1/markets/eu\",\"/v1/getboard\",\"/v1/board\",\"/v1/getticker\",\"/v1/ticker\",\"/v1/getexecutions\",\"/v1/executions\",\"/v1/getchats\",\"/v1/getchats/usa\",\"/v1/getchats/eu\",\"/v1/gethealth\",\"/v1/getboardstate\",\"/v1/me/getpermissions\",\"/v1/me/getbalance\",\"/v1/me/getcollateral\",\"/v1/me/getcollateralaccounts\",\"/v1/me/sendchildorder\",\"/v1/me/sendparentorder\",\"/v1/me/cancelchildorder\",\"/v1/me/cancelparentorder\",\"/v1/me/cancelallchildorders\",\"/v1/me/getchildorders\",\"/v1/me/getparentorders\",\"/v1/me/getparentorder\",\"/v1/me/getexecutions\",\"/v1/me/getpositions\",\"/v1/me/getcollateralhistory\",\"/v1/me/gettradingcommission\"]"

# Get Margin Status
get_collateral()
#> [1] "{\"collateral\":0.0,\"open_position_pnl\":0.0,\"require_collateral\":0.0,\"keep_rate\":0.0}"

#Send a New Order
x <- send_child_order(
  product_code = "BTC_JPY", 
  child_order_type = "LIMIT", 
  side = "BUY", 
  price = 50*10^4, 
  size = 0.001
)

# Cancel Order
id <- jsonlite::fromJSON(x)$child_order_acceptance_id
print(id)
#> [1] "JRF20180719-145022-924013"
cancel_child_order(product_code = "BTC_JPY", child_order_id = NULL, child_order_acceptance_id = id)
#> No encoding supplied: defaulting to UTF-8.
#> [1] ""

# Get Trading Commission
get_trading_commission(product_code = "BTC_JPY")
#> [1] "{\"commission_rate\":0.0015}"

# ... etc
```


NEWS.md
## bitflyer 0.1.0
- First release


tests/testthat.R
library(testthat)
library(bitflyer)

test_check("bitflyer")


tests/testthat/test-utils.R
context("utils.R")

test_that("get_from_env_or_global_env() works", {
  expect_null(get_from_env_or_global_env("HOGE"))
  # Global variable
  HOGE <- 123
  assign("HOGE", 123, envir=.GlobalEnv)
  expect_equal(get_from_env_or_global_env("HOGE"), 123)
  rm(HOGE, envir=.GlobalEnv)
  expect_null(get_from_env_or_global_env("HOGE"))
  # Environment variable
  Sys.setenv(HOGE = "ABC")
  expect_equal(get_from_env_or_global_env("HOGE"), "ABC")
  Sys.unsetenv("HOGE")
  expect_null(get_from_env_or_global_env("HOGE"))
})

test_that("calling_function_name() works", {
  f1 <- function(){calling_function_name(level=-1)}
  cf1 <- f1()
  expect_equal(cf1, "f1")
})


test_that("check_region() works", {
  # Japan
  for(region in c("", "jpn", "jp")){
    expect_equal(check_region(region), "")
  }

  # UAA
  for(region in c("usa", "us")){
    expect_equal(check_region(region), "usa")
  }

  # EURO
  for(region in c("eu", "euro")){
    expect_equal(check_region(region), "eu")
  }

  # Raise errors
  expect_error(check_region("chaina"))
  expect_error(check_region("japon"))
})


tests/testthat/test-public.R
context("public.R")

test_that("get_health()", {
  expect_equal(get_health(), "{\"status\":\"NORMAL\"}")
})


R/utils.R
get_from_env_or_global_env <- function(x)
{
  if(Sys.getenv(x) != ""){
    Sys.getenv(x)
  } else if(exists(x)){
    eval(parse(text = x), envir=.GlobalEnv)
  } else{
    NULL
  }
}

#' @noRd
calling_function_name <- function(level=-2)
{
  # e.g. bitflyer::get_markets (language)
  func_name <- as.list(sys.call(level))[[1]]
  # e.g. bitflyer::get_markets (character)
  func_name <- as.character(func_name)
  # e.g. get_markets (character)
  func_name <- func_name[length(func_name)]
  # e.g. getmarkets (character)
  stringr::str_replace_all(func_name, "_", "")
}

#' @noRd
build_path <- function(version, method, calling_function, region, query)
{
  if(length(query) >= 1 & method == "GET"){
    x <- purrr::imap(query, ~ paste0(.y, "=", .x))
    query <- paste0("?", purrr::reduce(x, ~ paste(.x, .y, sep = "&")))
  }
  region <- dplyr::if_else(region == "", "", paste0("/", region))
  paste0("/", version, "/", calling_function, region, query)
}

#' @noRd
build_url <- function(version, method, calling_function, region, query)
{
  path <- build_path(version, method, calling_function, region, query)
  paste0(BITFLYER_API_URL, path)
}

#' @noRd
check_region <- function(region) {
  region <- stringr::str_to_lower(region)
  japan <- c("", "jpn", "jp")
  usa <- c("usa", "us")
  euro <- c("eu", "euro")
  # Messy to use match.arg()
  # https://stackoverflow.com/questions/41441170/failure-of-match-arg-for-the-empty-string
  stopifnot(region %in% c(japan, usa, euro))
  for(country in list(japan, usa, euro)){
    if(region %in% country){
      return(country[1])
    }
  }
}

#' @noRd
stop_for_order_status <- function(status){
  ORDER_STATUSES <- c("COMPLETED", "CANCELED", "EXPIRED", "ACTIVE", "REJECTED")
  if(!is.null(status) && !(status %in% ORDER_STATUSES)){
    stop(sprintf("child_order_state must be one of %s", paste(ORDER_STATUSES, collapse = ", ")))
  }
}


#' @noRd
request <- function(method, url, ..., query = NULL, body = NULL)
{
  response <- if(method == "GET"){
    httr::GET(url, ..., query=query)
  } else{
    httr::POST(url, ..., body=body)
  }
  content <- httr::content(response, as="text")
  httr::stop_for_status(response, task = content)
  content
}


R/public.R
#' @include manual-template.R utils.R
NULL

# Functions in this souce file
request_public_get <- function(..., region = "")
{
  region <- check_region(region)
  name <- calling_function_name(-2)
  query <- list(...)
  url <- build_url("v1", "GET", name, region, query)
  request("GET", url, query = query)
}

# hoge <- func_product_code
# is equqal to the following code:
# hoge <- function(product_code = "BTC_JPY"){
#     request_public_get(product_code = product_code)
# }
#' @inheritParams region
func_product_code <- function(product_code = "BTC_JPY"){
  request_public_get(product_code = product_code)
}

func_region <- function(region = "") {
  request_public_get(region = region)
}

#' Market List
#'
#' Get the list of market
#'
#' @inheritParams region
#' @examples
#' \dontrun{
#'   markets()
#'   markets(region = "us")
#' }
#' @export
markets <- func_region

#' Order Book
#'
#' Get order book
#'
#' @inheritParams product_code
#' @examples
#' \dontrun{
#'   board()
#'   board(product_code = "BTC_USD")
#' }
#' @export
board <- func_product_code

#' Ticker
#'
#' Get ticker
#'
#' @inheritParams product_code
#' @examples
#' \dontrun{
#'   ticker(product_code = "BTC_JPY")
#' }
#' @export
ticker <- func_product_code

#' Execution History
#'
#' Get Execution History
#'
#' @inheritParams  product_code
#' @inheritParams  count_before_after
#' @examples
#' \dontrun{
#'   executions()
#'   executions(product_code="BCH_BTC", count=3, before="303218244")
#' }
#' @export
executions <- function(product_code = "BTC_JPY", count = 100, before = 0, after = 0){
  request_public_get(
    product_code = product_code,
    count = count,
    before = before,
    after = after
  )
}

#' Orderbook status
#'
#' Get orderbook status
#'
#' @inheritParams product_code
#' @examples
#' \dontrun{
#'   get_board_state(product_code = "BTC_JPY")
#' }
#'
#' @return status: one of the following levels will be displayed
#'   \itemize{
#'     \item NORMAL: The exchange is operating.
#'     \item BUSY: The exchange is experiencing high traffic.
#'     \item VERY BUSY: The exchange is experiencing very heavy traffic.
#'     \item SUPER BUSY: The exchange is experiencing extremely heavy traffic. There is a possibility that orders will fail or be processed after a delay.
#'     \item STOP: The exchange has been stopped. Orders will not be accepted.
#'  }
#'
#' @export
get_board_state <- func_product_code

#' Exchange status
#'
#' Get the current status of the exchange.
#'
#' @inheritParams product_code
#' @return
#'   One of the following levels
#'   \itemize{
#'     \item NORMAL: The exchange is operating.
#'     \item BUSY: The exchange is experiencing high traffic.
#'     \item VERY BUSY: The exchange is experiencing heavy traffic.
#'     \item SUPER BUSY: The exchange is experiencing extremely heavy traffic.
#'      There is a possibility that orders will fail or be processed after a delay.
#'     \item STOP: The exchange has been stopped. Orders will not be accepted.
#'   }
#' @examples
#' \dontrun{
#'   get_health()
#' }
#' @export
get_health <- func_product_code

#' Chat
#'
#' Get the history of chats
#'
#' @param from_date string (e.g. 2018-06-30).
#'   The function returns new messages after this date.
#' @examples
#' \dontrun{
#'   get_chats(from_date="2018-07-19")
#' }
#' @export
get_chats <- function(from_date){
  request_public_get(from_date = from_date)
}



R/private.R
#' @include manual-template.R utils.R
NULL

#' Functions in this souce file
#' @importFrom purrr compact
#' @noRd
request_private_get <- function(..., region = "")
{
  calling_function <- calling_function_name(-2)
  query <- purrr::compact(list(...))
  request_private(calling_function, "GET", query = query, region = region)
}

#' @noRd
request_private_post <- function(..., region = "")
{
  calling_function <- calling_function_name(-2)
  query <- purrr::compact(list(...))
  body <- jsonlite::toJSON(query, auto_unbox=TRUE)
  request_private(calling_function, "POST", body = body, region = region)
}

#' @noRd
request_private <- function(calling_function, method, query = NULL, body = NULL, region = "") {
  region <- check_region(region)
  path <- build_path("v1/me", method, calling_function, region, query)
  url <- build_url("v1/me", method, calling_function, region, query)
  # unix time stamp
  timestamp <- as.numeric(Sys.time())
  # Sign
  text <- paste0(timestamp, method, path, body)
  sign <- digest::hmac(key=get_from_env_or_global_env("BITFLYER_LIGHTNING_API_SECRET"), object=text, algo="sha256", serialize=FALSE)
  header <- httr::add_headers(
    `ACCESS-KEY`=get_from_env_or_global_env("BITFLYER_LIGHTNING_API_KEY"),
    `ACCESS-TIMESTAMP`=timestamp,
    `ACCESS-SIGN`=sign,
    `Content-Type`="application/json"
  )
  # Create header
  request(method, url, header, query = query, body = body)
}

#' @noRd
func_no_argument_private_get <- function() {
  request_private_get()
}

#' @noRd
func_product_code_private_get <- function(product_code) {
  request_private_get(product_code = product_code)
}

#' @noRd
func_count_before_after_private_get <- function(count = 100, before = 0, after = 0){
  request_private_get(count = count, before = before, after = after)
}


# Exported functions (APIs)

################################
# API
################################

#' Get API Key Permissions
#'
#' Get a list of which HTTP Private APIs can be used with the specified API key
#'
#' @export
get_permissions <- func_no_argument_private_get

################################
# Assets
################################

#' Get Account Asset Balance
#'
#' Get Account Asset Balance
#'
#' @export
get_balance <- func_no_argument_private_get

#' Get Margin Status
#'
#' Get Margin Status
#'
#' @export
get_collateral <- func_no_argument_private_get

#' Get Margin Status
#'
#' Get details of your collateral balances (multiple currencies supported).
#' @export
get_collateral_accounts <- func_no_argument_private_get

################################
# Trading
################################

#' Send a New Order
#'
#' Send a New Order
#'
#' @inheritParams product_code
#' @param child_order_type For limit orders, it will be "LIMIT". For market orders, "MARKET".
#' @param side For buy orders, "BUY". For sell orders, "SELL".
#' @param price Specify the price. This is a required value if child_order_type has been set to "LIMIT".
#' @param size Specify the order quantity.
#' @param minute_to_expire Specify the time in minutes until the expiration time.
#'   If omitted, the value will be 43200 (30 days).
#' @param time_in_force Specify any of the following execution conditions - "GTC", "IOC", or "FOK".
#'   If omitted, the value defaults to "GTC".
#'
#' @return child_order_acceptance_id.
#'   To specify the order to return, please use this instead of child_order_id.
#'   Please confirm the item is either Cancel Order or Obtain Execution List.
#'
#' @export
send_child_order <- function(product_code, child_order_type, side, price, size, minute_to_expire = 43200, time_in_force = "GTC") {
  request_private_post(
    product_code = product_code,
    child_order_type = child_order_type,
    side = side,
    price = price,
    size = size,
    minute_to_expire = minute_to_expire,
    time_in_force = time_in_force
  )
}

#' Cancel Order
#'
#' Cancel Order
#'
#' @inheritParams product_code
#' @inheritParams child_order_id
#' @inheritParams child_order_acceptance_id
#' @export
cancel_child_order <- function(product_code, child_order_id = NULL, child_order_acceptance_id = NULL) {
  request_private_post(
    product_code = product_code,
    child_order_id = child_order_id,
    child_order_acceptance_id = child_order_acceptance_id
  )
}

#' Submit New Parent Order (Special order)
#'
#' Submit New Parent Order (Special order)
#'
#' @details
#' It is possible to place orders including logic other than simple limit orders (LIMIT)
#' and market orders (MARKET). Such orders are handled as parent orders.
#' By using a special order, it is possible to place orders in response to market conditions or place multiple associated orders.
#'
#' Please read about the types of special orders and their methods in the bitFlyer Lightning documentation on special orders:
#' \url{https://lightning.bitflyer.com/docs/specialorder}.
#'
#' @param order_method The order method. Please set it to one of the following values. If omitted, the value defaults to "SIMPLE".
#'   \itemize{
#'     \item "SIMPLE": A special order whereby one order is placed.
#'     \item "IFD": Conducts an IFD order. In this method, you place two orders at once, and when the first order is completed, the second order is automatically placed.
#'     \item "OCO": Conducts an OCO order. In this method, you place two orders at one, and when one of the orders is completed, the other order is automatically canceled.
#'     \item "IFDOCO": Conducts an IFD-OCO order. In this method, once the first order is completed, an OCO order is automatically placed.
#'   }
#' @param minute_to_expire Specifies the time until the order expires in minutes. If omitted, the value defaults to 43200 (30 days).
#' @param time_in_force Specify any of the following execution conditions - "GTC", "IOC", or "FOK". If omitted, the value defaults to "GTC".
#' @param parameters This is an array that specifies the parameters of the order to be placed.
#'   The required length of the array varies depending upon the specified order_method.
#'   If "SIMPLE" has been specified, specify one parameter.
#'   If "IFD" has been specified, specify two parameters. The first parameter is the parameter for the first order placed. The second parameter is the parameter for the order to be placed after the first order is completed.
#'   If "OCO" has been specified, specify two parameters. Two orders are placed simultaneously based on these parameters.
#'   If "IFDOCO" has been specified, specify three parameters. The first parameter is the parameter for the first order placed. After the order is complete, an OCO order is placed with the second and third parameters.
#'   In the parameters, specify an array of objects with the following keys and values.
#' @inheritParams product_code
#' @param condition_type This is the execution condition for the order. Please set it to one of the following values.
#'   \itemize{
#'     \item "LIMIT": Limit order.
#'     \item "MARKET": Market order.
#'     \item "STOP": Stop order.
#'     \item "STOP_LIMIT": Stop-limit order.
#'     \item "TRAIL": Trailing stop order.
#'  }
#' @param side For buying orders, specify "BUY", for selling orders, specify "SELL".
#' @param size Specify the order quantity.
#' @param price Specify the price. This is a required value if condition_type has been set to "LIMIT" or "STOP_LIMIT".
#' @param trigger_price Specify the trigger price for a stop order. This is a required value if condition_type has been set to "STOP" or "STOP_LIMIT".
#' @param offset Specify the trail width of a trailing stop order as a positive integer. This is a required value if condition_type has been set to "TRAIL".
#'
#' @export
send_parent_order <- function(order_method, minute_to_expire, time_in_force, parameters, product_code, condition_type, side, size, price, trigger_price, offset) {
  request_private_post(
    order_method = order_method,
    minute_to_expire = minute_to_expire,
    time_in_force = time_in_force,
    parameters = parameters,
    product_code = product_code,
    condition_type = condition_type,
    side = side,
    size = size,
    price = price,
    trigger_price = trigger_price,
    offset = offset
  )
}

#' Cancel parent order
#'
#' Parent orders can be canceled in the same manner as regular orders.
#' If a parent order is canceled, the placed orders associated with that order will all be canceled.
#'
#' @inheritParams product_code
#' @inheritParams parent_order_id
#' @inheritParams parent_order_acceptance_id
#' @details
#' When \code{parent_order_acceptance_id} is specified, the corresponding order will be cancelled.
#' @export
cancel_parent_order <- request_private_post

#' Cancel All Orders
#'
#' Cancel All Orders
#'
#' @inheritParams product_code
#' @export
cancel_all_childorders <- func_product_code_private_get

#' List Orders
#'
#' List Orders
#'
#' @inheritParams product_code
#' @inheritParams count_before_after
#' @inheritParams child_order_state
#' @inheritParams child_order_id
#' @inheritParams child_order_acceptance_id
#' @inheritParams parent_order_id
#' @details
#' If \code{parent_order_id} is specified, a list of all orders associated with the parent order is obtained.
#' @importFrom glue glue
#' @export
get_child_orders <- function(product_code, count = 100, before = 0, after = 0, child_order_state = NULL, child_order_id = NULL, child_order_acceptance_id = NULL, parent_order_id = NULL){
  stop_for_order_status(child_order_state)
  request_private_get(
    product_code = product_code,
    count = count,
    before = before,
    after = after,
    child_order_state = child_order_state,
    child_order_id = child_order_id,
    child_order_acceptance_id = child_order_acceptance_id,
    parent_order_id = parent_order_id
  )
}

#' List Parent Orders
#'
#' List Parent Orders
#'
#' @inheritParams product_code
#' @inheritParams count_before_after
#' @inheritParams child_order_state
#' @export
get_parent_orders <- function(product_code, count = 100, before = 0, after = 0, child_order_state){
  request_private_get(
    product_code = product_code,
    count = count,
    before = before,
    after = after,
    child_order_state = child_order_state
  )
}

#' Get Parent Order Details
#'
#' Get Parent Order Details
#'
#' @inheritParams parent_order_id
#' @inheritParams parent_order_acceptance_id
#' @details
#' If \code{parent_order_acceptance_id} is specified, it returns the details of the parent order in question.
#' @export
get_parent_order <- function(parent_order_id, parent_order_acceptance_id){
    request_private_get(
      parent_order_id = parent_order_id,
      parent_order_acceptance_id = parent_order_acceptance_id
    )
}

#' List Executions
#'
#' List Executions
#'
#' @inheritParams product_code
#' @inheritParams count_before_after
#' @inheritParams child_order_id
#' @inheritParams child_order_acceptance_id
#' @details
#' When \code{child_order_id} is specified, a list of stipulations related to the order will be displayed.
#' When \code{child_order_acceptance_id} is specified, a list of stipulations related to the corresponding order will be displayed.
#' @export
get_executions <- function(product_code, count = 100, before = 0, after = 0, child_order_id = NULL, child_order_acceptance_id = NULL){
  request_private_get(
    product_code = product_code,
    count = count,
    before = before,
    after = after,
    child_order_id = child_order_id,
    child_order_acceptance_id = child_order_acceptance_id
  )
}

#' Get Open Interest Summary
#'
#' Get Open Interest Summary
#'
#' @param product_code Currently supports only "FX_BTC_JPY".
#' @export
get_positions <- function(){
  request_private_get(product_code = "FX_BTC_JPY")
}

#' Get Margin Change History
#'
#' Get Margin Change History
#'
#' @inheritParams count_before_after
#' @export
get_collateral_history <- func_count_before_after_private_get

#' Get Trading Commission
#'
#' Get Trading Commission
#'
#' @inheritParams product_code
#' @export
get_trading_commission <- func_product_code_private_get



# Not implemented yet.
################################
# Deposits and Withdrawals APIs
################################
# https://lightning.bitflyer.com/docs?lang=en#get-crypto-assets-deposit-addresses
# get_addresses
# get_coin_ins
# get_coin_outs
# get_bank_accounts
# get_deposits
# withdraw
# get_withdrawals



R/manual-template.R
#' @title product_code
#' @description product_code
#' @name product_code
#' @keywords internal
#' @param product_code A product_code or alias, as obtained from the Market List.
#'   If omitted, the value is set to "BTC_JPY".
#'   Only "BTC_USD" is available for U.S. accounts,
#'   and only "BTC_EUR" is available for European accounts.
NULL

#' @title region
#' @description region
#' @name region
#' @keywords internal
#' @param region Default value ("") is Japan.
#'   \itemize{
#'     \item Japan: ""(Default), "ja", "jpn
#'     \item Usa: "us", "usa"
#'     \item Euro: "eu"
#'   }
NULL

#' @title count, before, after
#' @description count_before_after
#' @name count_before_after
#' @keywords internal
#' @param count Specifies the number of results. If this is omitted, the value will be 100.
#' @param before Obtains data having an id lower than the value specified for this parameter.
#' @param after Obtains data having an id higher than the value specified for this parameter.
NULL


#' @title cild_order_state
#' @description cild_order_state
#' @name child_order_state
#' @keywords internal
#' @param child_order_state When specified, return only orders that match the specified value.
#'   You must specify one of the following:
#'   \itemize{
#'     \item "ACTIVE": Return open orders
#'     \item "COMPLETED": Return fully completed orders
#'     \item "CANCELED": Return orders that have been cancelled by the customer
#'     \item "EXPIRED": Return order that have been cancelled due to expiry
#'     \item "REJECTED": Return failed orders
#'   }
NULL


#' @title child_order_id
#' @description child_order_id
#' @name child_order_id
#' @keywords internal
#' @param child_order_id  ID for the child order.
NULL

#' @title child_order_acceptance_id
#' @description child_order_acceptance_id
#' @name child_order_acceptance_id
#' @keywords internal
#' @param child_order_acceptance_id Expects an ID from \code{\link{send_child_order}}
NULL

#' @title parent_order_id
#' @description child_order_acceptance_id
#' @name parent_order_id
#' @keywords internal
#' @param parent_order_id The ID of the parent order in question.
NULL

#' @title parent_order_acceptance_id
#' @description parent_order_acceptance_id
#' @name parent_order_acceptance_id
#' @keywords internal
#' @param parent_order_acceptance_id Expects an ID from \code{\link{send_parent_order}}
NULL


R/util.R
get_from_env_or_global_env <- function(x)
{
  if(Sys.getenv(x) != ""){
    Sys.getenv(x)
  } else if(exists(x)){
    eval(parse(text = x), envir=.GlobalEnv)
  }
}

calling_function_name <- function(level=-2)
{
  # e.g. bitflyer::get_markets (language)
  func_name <- as.list(sys.call(level))[[1]]
  # e.g. bitflyer::get_markets (character)
  func_name <- as.character(func_name)
  # e.g. get_markets (character)
  func_name <- func_name[length(func_name)]
  # e.g. getmarkets (character)
  stringr::str_replace_all(func_name, "_", "")
}

build_path <- function(version, method, calling_function, region, query)
{
  if(length(query) >= 1 & method == "GET"){
    x <- purrr::imap(query, ~ paste0(.y, "=", .x))
    query <- paste0("?", purrr::reduce(x, ~ paste(.x, .y, sep = "&")))
  }
  region <- dplyr::if_else(region == "", "", paste0("/", region))
  paste0("/", version, "/", calling_function, region, query)
}

build_url <- function(version, method, calling_function, region, query)
{
  path <- build_path(version, method, calling_function, region, query)
  paste0(BITFLYER_API_URL, path)
}

check_region <- function(region) {
  region <- stringr::str_to_lower(region)
  # Messy to use match.arg()
  # https://stackoverflow.com/questions/41441170/failure-of-match-arg-for-the-empty-string
  stopifnot(region %in% c("", "usa", "eu"))
  region
}

request <- function(method, url, ..., query = NULL, body = NULL)
{
  response <- if(method == "GET"){
    httr::GET(url, ..., query=query)
  } else{
    httr::POST(url, ..., body=body)
  }
  httr::stop_for_status(response)
  httr::content(response, as="text")
}

if(FALSE){
  get_health()
  get_markets()
  markets()
  markets(region="usa")
  board()
  get_ticker()
  ticker(product_code="BTC_USD")
  get_balance()
  #send_child_order(product_code="BTC_JPY", child_order_type="LIMIT", side="BUY", price=699000, size=0.001)
  get_collateral()
  get_child_orders(product_code="BTC_USD")
  get_child_orders(product_code="BTC_JPY")
  get_collateral_accounts()
  get_ticker()
  get_ticker(product_code="BTC_USD")
  # Private get
  get_trading_commission(product_code="BTC_JPY")
  get_trading_commission(product_code="ETH_BTC")
  # Private
  get_permissions()
  get_balance()
  get_collateral()
  get_collateral_accounts()
  get_addresses()
  get_coinins()
  get_coinouts()
  get_bank_accounts()
  get_deposits()
  withdraw()
  getwithdrawals()
  #Public
  markets()
  markets(region="USA")
  markets(region="eu")
  get_markets()
  get_markets(region="EU")
  board()
  board(product_code="BTC_JPY")
  board(product_code="BCH_BTC")
  get_board()
  get_board(product_code="BTC_JPY")
  get_board(product_code="BCH_BTC")
  ticker()
  ticker(product_code="BTC_JPY")
  ticker(product_code="BCH_BTC")
  get_ticker()
  get_ticker(product_code="BTC_JPY")
  get_ticker(product_code="BCH_BTC")
  executions()
  executions(product_code="BTC_JPY", count=3)
  executions(product_code="BCH_BTC", count=3, before="303218244")
  get_executions()
  get_health()
  get_health(product_code="BTC_JPY")
  get_health(product_code="BCH_BTC")
  get_chats()
}

make_request_private_get <- function(){}
make_request_private_post <- function(){}
